arch/x86_64/core/x86_64-hw.h

```diff
@@ -0,0 +1,282 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef _X86_64_HW_H
+#define _X86_64_HW_H
+
+/*
+ * Struct declarations and helper inlines for core x86_64 hardware
+ * functionality.  Anything related to ioports, CR's MSR's, I/L/GDTs,
+ * PTEs or (IO-)APICs can be found here.  Note that because this
+ * header is included in limited stub contexts, it should include
+ * declarations and inlines only: no data definitions, even extern
+ * ones!
+ */
+
+static inline unsigned long eflags(void)
+{
+	int eflags;
+
+	__asm__ volatile("pushfq; pop %%rax" : "=a"(eflags));
+	return eflags;
+}
+
+/* PAE page table record.  Note that "addr" is aligned naturally as an
+ * address, but of course must be masked to change only significant
+ * bits (which depend on whether it's storing a 4k, 2M or 1G memory
+ * block) so as to not clobber the bitfields (remember "negative"
+ * addresses must mask off the top bits too!).  The natural idiom is
+ * to assign addr first, then write the bitfields.
+ */
+struct pte64 {
+	union {
+		unsigned long long addr;
+		struct {
+			unsigned long long present : 1;
+			unsigned long long writable : 1;
+			unsigned long long usermode : 1;
+			unsigned long long writethrough : 1;
+			unsigned long long uncached : 1;
+			unsigned long long accessed : 1;
+			unsigned long long dirty : 1;
+			unsigned long long pagesize_pat : 1;
+			unsigned long long global : 1;
+			unsigned long long _UNUSED1 : 3;
+			unsigned long long pat : 1;
+			unsigned long long _UNUSED2 : 50;
+			unsigned long long exdisable : 1;
+		};
+	};
+};
+
+struct gdt64 {
+	union {
+		unsigned int dwords[2];
+		struct {
+			unsigned long long limit_lo16 : 16;
+			unsigned long long base_lo16 : 16;
+			unsigned long long base_mid8 : 8;
+			unsigned long long accessed : 1;
+			unsigned long long readable : 1;
+			unsigned long long conforming : 1;
+			unsigned long long codeseg : 1;
+			unsigned long long notsystem : 1;
+			unsigned long long ring : 2;
+			unsigned long long present : 1;
+			unsigned long long limit_hi4 : 4;
+			unsigned long long available : 1;
+			unsigned long long long64 : 1;
+			unsigned long long default_size : 1;
+			unsigned long long page_granularity : 1;
+			unsigned long long base_hi8 : 8;
+		};
+	};
+};
+
+static inline void gdt64_set_base(struct gdt64 *g, unsigned int base)
```

Comment:
suggest moving this and related defines to include/arch/x86/segmentation.h which we might itself move to some location shared between the two ports